### PR TITLE
feat: Revamp CrossValidationReport to use EstimatorReport

### PR DESCRIPTION
**Pull request by glemaitre**
closes #1014 

It is a revamp from the current `CrossValidationReporter`. I am working in another module such that:

- `CrossValidationReporter` is still available such that we do the transition for the UI part
- move `CrossValidationReport` to a private module since we want to encourage people to import either from `skore` or `skore.sklearn`.

### TODO

I have to do the TODO

### CODE REVIEW

File: skore/src/skore/utils/_progress_bar.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+    @classmethod
+    def stop_progress(cls):
+        if cls._progress is not None:
+            try:
+                cls._progress.stop()
+            finally:
+                cls._progress = None
+
+
+class ProgressDecorator:
```
Comment (augustebaum) [inline skore/src/skore/utils/_progress_bar.py:49]: This should be a function

Comment (glemaitre) [reply skore/src/skore/utils/_progress_bar.py:49]: @augustebaum you need to give more info.

FYI, I don't like the current design of the progress bar. I need to manually stop the thread manager and I need some private sentinel in the class to know if I should add a progress bar to another one to have nested progress bar. It is really invasive but it is a working POC.

So if you think that a design with a function can simplify and answer the different constraints and really looking forward. But you need to be explicit in how you see the design of this function, what problem does it solve compare to the actual design.

**Comment by augustebaum** [comment]
Side-note: Was it intentional to link to PR 538?

**Comment by glemaitre** [comment]
> Side-note: Was it intentional to link to PR 538?

Nop, it seems wrong. There might be a PR in the 500'ish to be close then :)

File: skore/src/skore/utils/_progress_bar.py (hunk 1 above)
Comment (glemaitre) [reply skore/src/skore/utils/_progress_bar.py:49]: > after all I'm reviewing a draft PR ^^

Always good to have feedback soonish ;)

> in its current state, a class containing nothing but a __call__ method is equivalent to a simple function, so it shouldn't be a class.

Yep I agree with this point. I think that the `description` can be passed with context manager as a function as well.

As said before, I don't like the design. I'm not sure that a decorator is great indeed.

Comment (glemaitre) [reply skore/src/skore/utils/_progress_bar.py]: Resolving since we made it better.

File: skore/src/skore/sklearn/_cross_validation/report.py [hunk 2]
```diff
@@ -0,0 +1,302 @@
...
+
+    y : array-like of shape (n_samples,) or (n_samples, n_outputs), default=None
+        The target variable to try to predict in the case of supervised learning.
+
+    cv_splitter : int, cross-validation generator or an iterable, default=5
+        Determines the cross-validation splitting strategy.
+        Possible inputs for `cv_splitter` are:
+
+        - int, to specify the number of folds in a `(Stratified)KFold`,
+        - a scikit-learn:term:`CV splitter`,
```
Comment (sylvaincom) [inline skore/src/skore/sklearn/_cross_validation/report.py:47]: I believe you are trying to provide a link to the scikit-learn glossary?

Comment (glemaitre) [reply skore/src/skore/sklearn/_cross_validation/report.py:47]: The linking works (at least locally) so I have good hope. Missing a space thought.

```suggestion
        - a scikit-learn :term:`CV splitter`,
```

File: skore/src/skore/sklearn/_cross_validation/report.py [hunk 3]
```diff
@@ -0,0 +1,302 @@
...
+        - int, to specify the number of folds in a `(Stratified)KFold`,
+        - a scikit-learn :term:`CV splitter`,
+        - An iterable yielding (train, test) splits as arrays of indices.
+
+        For int/None inputs, if the estimator is a classifier and ``y`` is
+        either binary or multiclass, :class:`StratifiedKFold` is used. In all
+        other cases, :class:`KFold` is used. These splitters are instantiated
+        with `shuffle=False` so the splits will be the same across calls.
+
+        Refer :ref:`User Guide <cross_validation>` for the various
```
Comment (sylvaincom) [inline skore/src/skore/sklearn/_cross_validation/report.py:55]:
```suggestion
        Refer to scikit-learn's :ref:`User Guide <cross_validation>` for the various
```

File: skore/src/skore/sklearn/_cross_validation/report.py [hunk 4]
```diff
@@ -0,0 +1,302 @@
...
+    )
+
+
+class CrossValidationReport(_BaseReport, DirNamesMixin):
+    """Reporter for cross-validation results.
+
+    Parameters
+    ----------
+    estimator : estimator object
+        Estimator to make report from.
```
Comment (sylvaincom) [inline skore/src/skore/sklearn/_cross_validation/report.py:34]:
```suggestion
        Estimator to make the cross-validation report from.
```

**Review by augustebaum — CHANGES_REQUESTED** [review]
Apart from the caching question I just have a few suggestions.
Also, I guess the feature where we save intermediate data (e.g. save after evaluating each split) is pushed back to later?

File: skore/src/skore/sklearn/_cross_validation/report.py [hunk 5]
```diff
@@ -0,0 +1,302 @@
...
+        >>> reporter.clear_cache()
+        >>> reporter._cache
+        {}
+        """
+        for report in self.estimator_reports_:
+            report.clear_cache()
+        self._cache = {}
+
+    @progress_decorator(description="Cross-validation predictions")
+    def cache_predictions(self, response_methods="auto", n_jobs=None):
```
[the remark this answers is not in the document]
Comment (glemaitre) [reply skore/src/skore/sklearn/_cross_validation/report.py:196]: Nop, because the `cache_predictions` are cached in each individual `EstimatorReport` under the hood. So we don't need to cache anything in the cache of the `CrossValidationReport`.

Trigger a metric and in this case, the aggregate metric is cached at the level of the `CrossValidationReport`.

**Comment by glemaitre** [comment]
Also, I guess the feature where we save intermediate data (e.g. save after evaluating each split) is pushed back to later?


So here, the data are stored in the `estimator_report` that is a generator, so we should be on the path. If the idea, is to make an update on the reporter shown in the activity feed, we would an interuption when consuming the reporter and trigger partial computation of the view. But we have all the primitive to do it I think

File: skore/src/skore/sklearn/_estimator/__init__.py [hunk 6]
```diff
@@ -11,13 +11,13 @@ def register_estimator_report_accessor(name: str):
     return _register_accessor(name, EstimatorReport)
 
 
-def register_metrics_accessor(name: str):
+def register_estimator_report_metrics_accessor(name: str):
     """Register an accessor for the EstimatorReport class."""
     return _register_accessor(name, _MetricsAccessor)
```
[the remark this answers is not in the document]
Comment (glemaitre) [reply skore/src/skore/sklearn/_estimator/__init__.py]: Done as well.

File: skore/src/skore/sklearn/_cross_validation/report.py (hunk 5 above)
[the remark this answers is not in the document]
Comment (glemaitre) [reply skore/src/skore/sklearn/_cross_validation/report.py:196]: To illustrate:

```python
from sklearn.datasets import make_classification
from sklearn.model_selection import train_test_split
from sklearn.linear_model import LogisticRegression
X, y = make_classification(random_state=42)
estimator = LogisticRegression()
from skore import CrossValidationReport
report = CrossValidationReport(estimator, X=X, y=y, cv_splitter=2)
report.cache_predictions()
# new call
report.metrics.accuracy()
report._cache
```

```
{(np.int64(-3970494336380670746),
  'accuracy',
  'test',
  None): Metric                       Accuracy (↗︎)
 LogisticRegression Split #0            1.0
                    Split #1            0.9}

```

**Comment by glemaitre** [comment]
The comment should be addressed @augustebaum @sylvaincom

File: .github/workflows/backend.yml [hunk 7]
```diff
@@ -102,10 +102,12 @@ jobs:
         run: |
```
Comment (thomass-dev) [inline .github/workflows/backend.yml:102]: something wrong with the rebase/merge commit?

Comment (glemaitre) [reply .github/workflows/backend.yml]: Yep apparently. Should be solved

**Review by augustebaum — APPROVED** [review]
Many thanks
